[SageMaker] Add GraphStorm SageMaker Pipeline creation and execution (#1108)
awslabs/graphstorm: sagemaker/pipeline/README.md

## Installation

To construct and execute GraphStorm SageMaker pipelines you need the code
available and a Python environment with the SageMaker SDK and `boto3` installed.

1. Clone the GraphStorm repository:
   ```
   git clone https://github.com/awslabs/graphstorm.git
   cd graphstorm/sagemaker/pipeline
   ```

2. Install the required dependencies:
   ```
   pip install sagemaker boto3
   ```

## Usage

### Creating a Pipeline

To create a new SageMaker pipeline for GraphStorm:

```bash
python create_sm_pipeline.py \
    --execution-role arn:aws:iam::123456789012:role/SageMakerRole \
    --region us-west-2 \
    --graphstorm-pytorch-cpu-image-url 123456789012.dkr.ecr.us-west-2.amazonaws.com/graphstorm:sagemaker-cpu \
    --instance-count 2 \
    --jobs-to-run gconstruct train inference \
    --graph-name my-graph \
    --graph-construction-config-filename my_gconstruct_config.json \
    --input-data-s3 s3://input-bucket/data \
    --output-prefix s3://output-bucket/results \
    --train-inference-task node_classification \
    --train-yaml-s3 s3://config-bucket/train.yaml
```

This command creates a new pipeline with the specified configuration. The pipeline will

> include one GConstruct job, one training job and one inference job.

Reviewer: You also need to introduce the `execution-role` and `graphstorm-pytorch-cpu-image-url` which are very important.

It will use the configuration defined in `s3://input-bucket/data/my_gconstruct_config.json`
to construct the graph and the train config file at `s3://config-bucket/train.yaml`
to run training and inference.

The `--instance-count` parameter determines the number of workers and partitions we will create and use
during partitioning/training.

You can customize various aspects of the pipeline using additional command-line arguments. Refer to the script's help message for a full list of options:

```bash
python create_sm_pipeline.py --help
```

### Executing a Pipeline

To execute a created pipeline:

```bash

> python execute_sm_pipeline.py \

Reviewer: Have you defined the pipeline-name in L68?

    --pipeline-name my-graphstorm-pipeline \
    --region us-west-2
```

> You can override various pipeline parameters during execution:

Reviewer: ```suggestion
You can override the default pipeline parameters during execution:
```

```bash
python execute_sm_pipeline.py \
    --pipeline-name my-graphstorm-pipeline \
    --region us-west-2 \
    --instance-count 4 \
    --gpu-instance-type ml.g4dn.12xlarge
```

For a full list of execution options:

```bash
python execute_sm_pipeline.py --help
```

> For more fine-grained execution options, like selective execution,

Reviewer: ```suggestion
For more fine-grained execution options, like selective execution, please refer to 
```

[SageMaker AI documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/pipelines-selective-ex.html).

## Pipeline Components

The GraphStorm SageMaker pipeline can include the following steps:

1. **Graph Construction (GConstruct)**: Builds the partitioned graph from input data in a single instance.

> 2. **Graph Processing (GSProcessing)**: Processes the graph data using PySpark, preparing it for distributed partitioning.

Reviewer: ```suggestion
2. **Graph Processing (GSProcessing)**: Processes the graph data using PySpark, preparing it for distributed graph partitioning.
```

3. **Graph Partitioning (DistPart)**: Partitions the graph using multiple instances.
4. **GraphBolt Conversion**: Converts the partitioned data (usually generated from DistPart) to GraphBolt format.
5. **Training**: Trains the graph neural network model.
6. **Inference**: Runs inference on the trained model.

Each step is configurable and can be customized based on your specific requirements.

## Configuration

The pipeline's behavior is controlled by various configuration parameters, including:

- AWS configuration (region, roles, image URLs)
- Instance configuration (instance types, counts)
- Task configuration (graph name, input/output locations)
- Training and inference configurations

> Refer to the `PipelineArgs` class in `pipeline_parameters.py` for a complete list of configurable options.

Reviewer: We need to list the options here when we officially announce the support of SageMaker pipline.

## Advanced Usage

### Using GraphBolt

To use GraphBolt for improved performance:

```bash
python create_sm_pipeline.py \
    ... \
    --use-graphbolt true
```

### Custom Job Sequences

You can customize the sequence of jobs in the pipeline using the `--jobs-to-run` argument when creating the pipeline. For example:

```bash
python create_sm_pipeline.py \
    ... \
    --jobs-to-run gsprocessing dist_part gb_convert train inference \
    --use-graphbolt true
```

This will create a pipeline that uses GSProcessing to process and prepare the data for partitioning,

> use DistPart to partition the data, convert the partitioned data to the GraphBolt format,

Reviewer: ```suggestion
uses DistPart to partition the data, converts the partitioned data to the GraphBolt format,
```

> then run a train and an inference job in sequence.

Reviewer: ```suggestion
then runs a train and an inference job in sequence.
```

You can use this job sequence when your graph is too large to partition on one instance using

> GConstruct (1+ TB is the suggested threshold to move to distributed partitioning).

Reviewer: 1TB+ of what? edges or nodes?

- `--graphstorm-pytorch-gpu-image-url`: GraphStorm GConstruct/dist_part/train/inference GPU ECR image URL.
- `--gsprocessing-pyspark-image-url`: GSProcessing SageMaker PySpark ECR image URL.

### Instance Configuration
- `--instance-count` / `--num-parts`: Number of worker instances/partitions for partition, training, inference. (Required)
- `--cpu-instance-type`: CPU instance type. (Default: ml.m5.4xlarge)
- `--gpu-instance-type`: GPU instance type. (Default: ml.g5.4xlarge)
- `--train-on-cpu`: Run training and inference on CPU instances instead of GPU. (Flag)
- `--graph-construction-instance-type`: Instance type for graph construction.
- `--gsprocessing-instance-count`: Number of GSProcessing instances.
- `--volume-size-gb`: Additional volume size for SageMaker instances in GB. (Default: 100)

### Task Configuration
- `--graph-name`: Name of the graph. (Required)
- `--input-data-s3`: S3 path to the input graph data. (Required)
- `--output-prefix-s3`: S3 prefix for the output data. (Required)
- `--pipeline-name`: Name for the pipeline.
- `--base-job-name`: Base job name for SageMaker jobs. (Default: 'gs')

> - `--jobs-to-run`: Space-separated string of jobs to run in the pipeline. (Required)

Reviewer: Is there any place, a user can find the names of jobs supported in the pipeline?

Author: Good point, currently they need to run `--help` in the creation, I'll document them here as well.

- `--log-level`: Logging level for the jobs. (Default: INFO)
- `--step-cache-expiration`: Expiration time for the step cache. (Default: 30d)
- `--update-pipeline`: Update an existing pipeline instead of creating a new one. (Flag)

### Graph Construction Configuration
- `--graph-construction-config-filename`: Filename for the graph construction config.
- `--graph-construction-args`: Parameters to be passed directly to the GConstruct job.

### Partition Configuration
- `--partition-algorithm`: Partitioning algorithm. (Default: random)
- `--partition-output-json`: Name for the output JSON file that describes the partitioned data. (Default: metadata.json)
- `--partition-input-json`: Name for the JSON file that describes the input data for partitioning. (Default: updated_row_counts_metadata.json)

### Training Configuration
- `--model-output-path`: S3 path for model output.
- `--num-trainers`: Number of trainers to use during training/inference. (Default: 4)

> - `--train-inference-task`: Task type for training and inference. (Required)

Reviewer: ```suggestion
- `--train-inference-task-type`: Task type for training and inference. (Required)
```
The name train-inference-task is confusing.

```

This will create a pipeline that uses GSProcessing to process and prepare the data for partitioning,
uses DistPart to partition the data, converts the partitioned data to the GraphBolt format,
then runs a train and an inference job in sequence.
You can use this job sequence when your graph is too large to partition on one instance using
GConstruct. 10B+ edges is the suggested threshold to move to distributed partitioning, or if your

> features are larger than 1TB.

Reviewer: ```suggestion
features are larger than 1TByte.
```
